design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,151 @@
+    // ...
+}
+```
+
+```cairo
+#[test]
+fn my_test_1() {
+    let a = MyStruct { a: 1, b: array![2, 3] };
+    // ...
+}
+```
+
+### Problems With Code Generation
+
+The code generation solution has some problems that will need to be addressed before it is implemented.
+See [section below](#possible-problems-with-code-generation) for details.
+
+## User Experience
+
+Unlike fuzz-tests, we should run and indicate the result of all parametrized cases.
+If some test cases fail, others should still be executed and their results should be displayed to the user.
+
+An example output could look similarly to this:
+
+```shell
+$ snforge test
+[PASS] tests::parametrized(a = 1, b = 2)
+[PASS] tests::parametrized(a = 3, b = 5)
+[FAIL] tests::parametrized(a = 4, b = 5)
+
+Failure data:
+    original value: [344693033283], converted to a string: [PANIC]
+    
+[PASS] tests::parametrized(a = 5, b = 7)
+# ...
+```
+
+## Deterministic Test Output Order
+
+Parallel test execution implementation in forge makes deterministic test execution order non-trivial.
+For good user experience, we should aim to implement parametrized tests in such a way that cases results are displayed
+in deterministic order and grouped together.
+
+Test execution itself can be performed in any order as long as we display the results in a grouped manner.
+We can follow the fuzzed tests implementation for this.
+
+### Behavior of `--exit-first` Flag
+
+In case `--exit-first` flag is used we should try to have consistent behavior of parametrized tests.
+
+If some other test fails and there are cases of parametrized tests that did not finish execution,
+we should not display any results of parametrized test.
+
+In case a parametrized test case fails, we should display it.
+Other cases may not be displayed in this case.
+
+## Required Changes to Test Collector
+
+To support parametric test, we need to introduce some changes to the test collector in Scarb as outlined below.
+
+### Define the Necessary Attributes
+
+This change is mostly straightforward: It can be done in the same manner as all other attributes.
+
+### Code Generation
+
+For this approach, my recommendation is to use plugins for code generation.
+We could create a separate plugin for just generating test cases from our attribute and make sure it is executed before
+the test collection.
+
+Cairo repository contains multiple examples of plugins that generate Cairo code which we could follow:
+
+- https://github.com/starkware-libs/cairo/blob/main/crates/cairo-lang-plugins/src/plugins/generate_trait.rs
+- https://github.com/starkware-libs/cairo/blob/main/crates/cairo-lang-plugins/src/plugins/panicable.rs
+- https://github.com/starkware-libs/cairo/blob/main/crates/cairo-lang-semantic/src/inline_macros/array.rs
+
+Details of this are for the implementer to investigate, but my recommendation would be to not attempt any approaches
+where we try to modify the test code in the compiler itself but generate the necessary test cases.
+
```

Comment:
Can we have at the end some takeaways and a stage-based plan for implementation/mvp?

Reply:
Wait, actually we want to generate the code in the compiler and then inject it into the db, right? Isn't it what the plugins do? I may be wrong tho

Reply:
What I meant by this is we want to just generate cairo code and inject it. Avoid any kind of magic where we try to modify already parsed code (if that's even possible).

Reply:
Could we make it more clear then? I understood it as the opposite